### PR TITLE
Telemetry: Don't attempt to compute route when subresources are used

### PR DESCRIPTION
**Pull request by jakub-pomykala**
~Still a work in progress. Tests are running but failing due to no current solution for getting the correct url when using sub resources.~

Changes in the `TelemetryContainerFilter` class move the logic for getting `route` into a separate method and also add in a check for the list size so that if there are any sub resources being use the `route` will not be modified.

For #25058

### CODE REVIEW

File: dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/jaxrspropagation/route/JaxRsRouteTestServlet.java [hunk 1]
```diff
@@ -111,6 +111,70 @@ public void testRouteWithQueryParam() {
...
+                        .withAttribute(SemanticAttributes.HTTP_METHOD, "GET")
+                        .withAttribute(SemanticAttributes.HTTP_STATUS_CODE, 200L)
+                        .withAttribute(SemanticAttributes.HTTP_URL, testUri.toString() + "getSubResourceWithPathParam/myIdForTesting/details"));
+
+        assertThat(serverSpan, isSpan()
+                        .withKind(SpanKind.SERVER)
+                        .withAttribute(SemanticAttributes.HTTP_METHOD, "GET")
+                        .withAttribute(SemanticAttributes.HTTP_STATUS_CODE, 200L)
+                        .withAttribute(SemanticAttributes.HTTP_ROUTE, getPath() + "getSubResourceWithPathParam/{id}/details")
+                        .withAttribute(SemanticAttributes.HTTP_TARGET, getPath() + "getSubResourceWithPathParam/{id}/details"));
```
Comment (Azquelt) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/jaxrspropagation/route/JaxRsRouteTestServlet.java]:
```suggestion
                        .withAttribute(SemanticAttributes.HTTP_TARGET, getPath() + "getSubResourceWithPathParam/myIdForTesting/details"));
```

Comment (jakub-pomykala) [reply dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/jaxrspropagation/route/JaxRsRouteTestServlet.java]: Done

Comment (Azquelt) [reply dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/jaxrspropagation/route/JaxRsRouteTestServlet.java]: Looks like this change has got lost somewhere

File: dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/jaxrspropagation/route/JaxRsRouteTestServlet.java [hunk 2]
```diff
@@ -111,6 +111,70 @@ public void testRouteWithQueryParam() {
...
+                        .withKind(SpanKind.CLIENT)
+                        .withAttribute(SemanticAttributes.HTTP_METHOD, "GET")
+                        .withAttribute(SemanticAttributes.HTTP_STATUS_CODE, 200L)
+                        .withAttribute(SemanticAttributes.HTTP_URL, testUri.toString() + "/getSubResourceWithQueryParam?id=myIdForTesting"));
+
+        assertThat(serverSpan, isSpan()
+                        .withKind(SpanKind.SERVER)
+                        .withAttribute(SemanticAttributes.HTTP_METHOD, "GET")
+                        .withAttribute(SemanticAttributes.HTTP_STATUS_CODE, 200L)
+                        .withAttribute(SemanticAttributes.HTTP_ROUTE, getPath() + "/getSubResourceWithQueryParam"));
```
Comment (Azquelt) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/jaxrspropagation/route/JaxRsRouteTestServlet.java]: Please test `HTTP_TARGET` here as well

File: dev/io.openliberty.microprofile.telemetry.1.1.internal/src/io/openliberty/microprofile/telemetry11/internal/rest/TelemetryContainerFilter.java [hunk 3]
```diff
@@ -145,34 +145,22 @@ public void filter(final ContainerRequestContext request) {
...
-                    route = template.toTemplate();
-                    ROUTE_CACHE.putRoute(resourceClass, resourceMethod, route);
+                if (route != null) {
+                    currentSpan.setAttribute(SemanticAttributes.HTTP_ROUTE, route);
+                    currentSpan.updateName(request.getMethod() + " " + route);
                 }
 
                 currentSpan.setAttribute(SemanticAttributes.HTTP_ROUTE, route);
-                currentSpan.updateName(request.getMethod()  + " " + route);
+                currentSpan.updateName(request.getMethod() + " " + route);
```
Comment (Azquelt) [inline dev/io.openliberty.microprofile.telemetry.1.1.internal/src/io/openliberty/microprofile/telemetry11/internal/rest/TelemetryContainerFilter.java]: You're setting the attribute and the name both inside and outside of your if block

File: dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry10/internal/rest/TelemetryContainerFilter.java [hunk 4]
```diff
@@ -208,6 +195,36 @@ public void filter(final ContainerRequestContext request, final ContainerRespons
...
+
+        String route = ROUTE_CACHE.getRoute(resourceClass, resourceMethod);
+
+        if (route == null) {
+
+            int checkResourceSize = request.getUriInfo().getMatchedResources().size();
+
+            // Check the resource size using getMatchedResource()
+            // A resource size > 1 indicates that there is a subresource
+            if (checkResourceSize == 1) {
```
Comment (Azquelt) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry10/internal/rest/TelemetryContainerFilter.java]:
```suggestion
            int resourceCount = request.getUriInfo().getMatchedResources().size();

            // Check the resource count using getMatchedResource()
            // A resource count > 1 indicates that there is a subresource
            // We can't currently compute the route correctly when subresources are used
            if (resourceCount == 1) {
```

Please make this change in the 1.1 version of `TelemetryContainerFilter` as well

File: dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry10/internal/rest/TelemetryContainerFilter.java [hunk 5]
```diff
@@ -153,30 +153,17 @@ public void filter(final ContainerRequestContext request) {
...
                     }
-
-                    currentSpan.setAttribute(SemanticAttributes.HTTP_ROUTE, route);
-                    currentSpan.updateName(route);
                 }
             }
-        } catch (Exception e) {
+        } catch (
+
+        Exception e) {
```
Comment (Azquelt) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry10/internal/rest/TelemetryContainerFilter.java]: Extra newlines here

**Comment by jakub-pomykala** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.microprofile.telemetry.1.0.internal_fat,io.openliberty.microprofile.telemetry.1.0.internal.jaeger_fat,io.openliberty.microprofile.telemetry.1.0.internal_fat_tck,io.openliberty.microprofile.telemetry.1.1.internal_fat_tck

**Review by Azquelt — APPROVED** [review]

**Comment by Azquelt** [comment]
#libby

**Pull request merged**